src/mscorlib/src/System/Diagnostics/Eventing/EventPipe.cs

```diff
@@ -0,0 +1,157 @@
+            m_providers = new List<EventPipeProviderConfiguration>();
+        }
+
+        internal string OutputFile
+        {
+            get { return m_outputFile; }
+        }
+
+        internal uint CircularBufferSizeInMB
+        {
+            get { return m_circularBufferSizeInMB; }
+        }
+
+        internal EventPipeProviderConfiguration[] Providers
+        {
+            get { return m_providers.ToArray(); }
+        }
+
+        internal void EnableProvider(string providerName, UInt64 keywords, uint loggingLevel)
+        {
+            m_providers.Add(new EventPipeProviderConfiguration(
+                providerName,
+                keywords,
+                loggingLevel));
+        }
+    }
+
+    internal static class EventPipe
+    {
+        internal static void Enable(EventPipeConfiguration configuration)
+        {
+            if(configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            EventPipeProviderConfiguration[] providers = configuration.Providers;
+
+            EventPipeInternal.Enable(
+                configuration.OutputFile,
+                configuration.CircularBufferSizeInMB,
+                providers,
+                providers.Length);
+        }
+
+        internal static void Disable()
+        {
+            EventPipeInternal.Disable();
+        }
+    }
+
+    internal static class EventPipeInternal
+    {
+        //
+        // These PInvokes are used by the configuration APIs to interact with EventPipe.
+        //
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        [SuppressUnmanagedCodeSecurity]
+        internal static extern void Enable(string outputFile, uint circularBufferSizeInMB, EventPipeProviderConfiguration[] providers, int numProviders);
+
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        [SuppressUnmanagedCodeSecurity]
+        internal static extern void Disable();
+
+        //
+        // These PInvokes are used by EventSource to interact with the EventPipe.
+        //
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        [SuppressUnmanagedCodeSecurity]
+        internal static extern IntPtr CreateProvider(Guid providerID, UnsafeNativeMethods.ManifestEtw.EtwEnableCallback callbackFunc);
+
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        [SuppressUnmanagedCodeSecurity]
+        internal static extern IntPtr AddEvent(IntPtr provHandle, Int64 keywords, uint eventID, uint eventVersion, uint level, bool needStack);
+
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        [SuppressUnmanagedCodeSecurity]
+        internal static extern void DeleteProvider(IntPtr provHandle);
+
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        [SuppressUnmanagedCodeSecurity]
+        internal static extern unsafe void WriteEvent(IntPtr eventHandle, void* data, uint length);
```

Comment:
One of the major cleanup items is to fix an issue with custom filtering and ETW (bug 105200).  TO fix this we need to be able NOT write to every session with a particular WriteEvent call.  We will be doing this by using the Filter parameter of the [EventWriteEx](https://msdn.microsoft.com/en-us/library/windows/desktop/dd392307(v=vs.85).aspx) method, and we will need to plumb that through here as well.      I debated asking you to do this plumbing now, but It is probably better to wait.   

I do want you to be aware that ultimately however, we need a parameter passed to WriteEvent that represents the set of sessions to send the event to (which we don't today, and we default to 'all.